libs/providers/multi-provider/src/lib/multi-provider.ts

```diff
@@ -0,0 +1,325 @@
+export type RegisteredProvider = {
+  name: string;
+  provider: Provider;
+};
+
+export class MultiProvider implements Provider {
+  readonly runsOn = 'server';
+
+  public readonly events = new OpenFeatureEventEmitter();
+
+  private hookContexts: WeakMap<EvaluationContext, HookContext> = new WeakMap<EvaluationContext, HookContext>();
+  private hookHints: WeakMap<EvaluationContext, HookHints> = new WeakMap<EvaluationContext, HookHints>();
+
+  metadata: ProviderMetadata;
+
+  providerEntries: RegisteredProvider[] = [];
+  private providerEntriesByName: Record<string, RegisteredProvider> = {};
+
+  private hookExecutor: HookExecutor;
+  private statusTracker = new StatusTracker(this.events);
+
+  constructor(
+    readonly constructorProviders: ProviderEntryInput[],
+    private readonly evaluationStrategy: BaseEvaluationStrategy = new FirstMatchStrategy(),
+    private readonly logger: Logger = new DefaultLogger(),
+  ) {
+    this.hookExecutor = new HookExecutor(this.logger);
+
+    this.registerProviders(constructorProviders);
+
+    const aggregateMetadata = Object.keys(this.providerEntriesByName).reduce((acc, name) => {
+      return { ...acc, [name]: this.providerEntriesByName[name].provider.metadata };
+    }, {});
+
+    this.metadata = {
+      ...aggregateMetadata,
+      name: MultiProvider.name,
+    };
+  }
+
+  private registerProviders(constructorProviders: ProviderEntryInput[]) {
+    const providersByName: Record<string, Provider[]> = {};
+
+    for (const constructorProvider of constructorProviders) {
+      const providerName = constructorProvider.provider.metadata.name;
+      let candidateName = constructorProvider.name ?? providerName;
+
+      if (constructorProvider.name && providersByName[constructorProvider.name]) {
+        throw new Error('Provider names must be unique');
+      }
+
+      providersByName[candidateName] ||= [];
+      providersByName[candidateName].push(constructorProvider.provider);
+    }
+
+    for (const name of Object.keys(providersByName)) {
+      const useIndexedNames = providersByName[name].length > 1;
+      for (let i = 0; i < providersByName[name].length; i++) {
+        const indexedName = useIndexedNames ? `${name}-${i + 1}` : name;
+        this.providerEntriesByName[indexedName] = { provider: providersByName[name][i], name: indexedName };
+        this.providerEntries.push(this.providerEntriesByName[indexedName]);
+        this.statusTracker.wrapEventHandler(this.providerEntriesByName[indexedName]);
+      }
+    }
+
+    // just make sure we don't accidentally modify these later
+    Object.freeze(this.providerEntries);
+    Object.freeze(this.providerEntriesByName);
+  }
+
+  async initialize(context?: EvaluationContext): Promise<void> {
+    await Promise.all(this.providerEntries.map((provider) => provider.provider.initialize?.(context)));
+  }
+
+  async onClose(): Promise<void> {
+    await Promise.all(this.providerEntries.map((provider) => provider.provider.onClose?.()));
```

Comment:
Maybe we can do `Promise.allSettled` here? This way every provider will be closed, even if one fails to do so.

Reply:
updated to use `allSettled` for both "onClose" and "initialize". It will also construct an "aggregate error" out of the error results from each provider

Reply:
Great, just added a comment regarding the AllSettled type :)